File: src/main/scala/com/xmlcalabash/runtime/ProcessMatchingNodes.scala
package com.xmlcalabash.runtime

import net.sf.saxon.om.AttributeMap
import net.sf.saxon.s9api.XdmNode

trait ProcessMatchingNodes {
  def startDocument(node: XdmNode): Boolean
  def endDocument(node: XdmNode): Unit

  def startElement(node: XdmNode, attribute: AttributeMap): Boolean
  def attributes(node: XdmNode, matchingAttributes: AttributeMap, nonMatchingAttributes: AttributeMap): Option[AttributeMap]
  def endElement(node: XdmNode): Unit

  def text(node: XdmNode): Unit
  def comment(node: XdmNode): Unit
  def pi(node: XdmNode): Unit
}

File: xmlcalabash/src/main/kotlin/com/xmlcalabash/runtime/ProcessMatchingNodes.kt
package com.xmlcalabash.runtime

import net.sf.saxon.om.AttributeMap
import net.sf.saxon.s9api.XdmNode

interface ProcessMatchingNodes {
    fun startDocument(node: XdmNode): Boolean
    fun endDocument(node: XdmNode)

    fun startElement(node: XdmNode, attributes: AttributeMap): Boolean
    fun attributes(node: XdmNode, matchingAttributes: AttributeMap, nonMatchingAttributes: AttributeMap): AttributeMap?
    fun endElement(node: XdmNode)

    fun text(node: XdmNode)
    fun comment(node: XdmNode)
    fun pi(node: XdmNode)
}